src/main/java/com/zendesk/maxwell/MaxwellHA.java

```diff
@@ -0,0 +1,67 @@
+package com.zendesk.maxwell;
+
+import org.jgroups.JChannel;
+import org.jgroups.protocols.raft.RaftLeaderException;
+import org.jgroups.protocols.raft.Role;
+import org.jgroups.protocols.raft.StateMachine;
+import org.jgroups.raft.RaftHandle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class MaxwellHA {
+	static final Logger LOGGER = LoggerFactory.getLogger(MaxwellHA.class);
+
+	private final Maxwell maxwell;
+	private final String jgroupsConf, raftMemberID, clientID;
+	private boolean hasRun = false;
+	private AtomicBoolean isRaftLeader = new AtomicBoolean(false);
+
+	public MaxwellHA(Maxwell maxwell, String jgroupsConf, String raftMemberID, String clientID) {
+		this.maxwell = maxwell;
+		this.jgroupsConf = jgroupsConf;
+		this.raftMemberID = raftMemberID;
+		this.clientID = clientID;
+	}
+
+	public void startHA() throws Exception {
+		JChannel ch=new JChannel(jgroupsConf);
+		RaftHandle handle=new RaftHandle(ch, null);
+		if ( raftMemberID != null )
+			handle.raftId(raftMemberID);
+		else
+			LOGGER.warn("--raft_member_id not specified, using values from " + jgroupsConf);
+
+		handle.addRoleListener(role -> {
+			if(role == Role.Leader) {
+				LOGGER.info("won HA election, starting maxwell");
+				isRaftLeader.set(true);
+
+				if ( hasRun )
+					maxwell.restart();
+				else {
+					maxwell.run();
+					hasRun = true;
+				}
+
+				isRaftLeader.set(false);
+			} else if ( this.isRaftLeader.get() ) {
+				LOGGER.info("Unable to find consensus, stepping down HA leadership");
+				maxwell.terminate();
+				isRaftLeader.set(false);
+			} else {
+				LOGGER.info("lost HA election, current leader: " + handle.leader());
+			}
+		});
+
+		ch.connect(this.clientID);
+		LOGGER.info("enter HA group, current leader: " +  handle.leader());
+
+		Thread.sleep(Long.MAX_VALUE);
```

Comment:
If `maxwell.run()` fails to execute, I'm assuming in non HA mode the process dies.  But here it looks like the process would just hang.  Is that right?

Reply:
very good question, lemme study the code and/or run it and figure out the answer to this one.  I don't think we should attempt to continue if `maxwell.run()` crashes for whatever reason.

Reply:
utility-of-code-review of the year award there @lukestephenson-zendesk.  Updated so that HA process still crash on exception.